packages/event-handler/src/rest/errors.ts

```diff
@@ -15,3 +18,145 @@ export class ParameterValidationError extends RouteMatchingError {
     this.name = 'ParameterValidationError';
   }
 }
+
+abstract class ServiceError extends Error {
+  abstract readonly statusCode: number;
+  abstract readonly errorType: string;
+  public readonly details?: Record<string, unknown>;
+
+  constructor(
+    message: string,
+    options?: ErrorOptions,
+    details?: Record<string, unknown>
+  ) {
+    super(message, options);
+    this.name = 'ServiceError';
+    this.details = details;
+  }
+
+  toJSON(): ErrorResponse {
+    return {
+      statusCode: this.statusCode,
+      error: this.errorType,
+      message: this.message,
+      ...(this.details && { details: this.details }),
+    };
+  }
+}
+
+export class BadRequestError extends ServiceError {
+  readonly statusCode = HttpErrorCodes.BAD_REQUEST;
+  readonly errorType = 'BadRequestError';
+
+  constructor(
+    message?: string,
+    options?: ErrorOptions,
+    details?: Record<string, unknown>
+  ) {
+    super(message ?? 'Bad request', options, details);
+  }
+}
+
+export class UnauthorizedError extends ServiceError {
+  readonly statusCode = HttpErrorCodes.UNAUTHORIZED;
+  readonly errorType = 'UnauthorizedError';
+
+  constructor(
+    message?: string,
+    options?: ErrorOptions,
+    details?: Record<string, unknown>
+  ) {
+    super(message ?? 'Unauthorized', options, details);
+  }
+}
+
+export class ForbiddenError extends ServiceError {
+  readonly statusCode = HttpErrorCodes.FORBIDDEN;
+  readonly errorType = 'ForbiddenError';
+
+  constructor(
+    message?: string,
+    options?: ErrorOptions,
+    details?: Record<string, unknown>
+  ) {
+    super(message ?? 'Forbidden', options, details);
```

Comment:
This is what I mean by having to add dummy values here rather than making `message` optional on the base class and just passing it through even if it's `undefined`.

Reply:
What about this @dreamorosi ? Should we make `message` optional in the superclass?